fix(deps): request consumers to install react-native polyfills

trivikr opened the pull request:
Co-authored-by: Nicolas Perraut <n.perraut@gmail.com>

### Issue
Related to #1536
Replaces https://github.com/aws/aws-sdk-js-v3/pull/2108

Fixes: https://github.com/aws/aws-sdk-js-v3/issues/1797
Fixes: https://github.com/aws/aws-sdk-js-v3/issues/2051
Fixes: https://github.com/aws/aws-sdk-js-v3/issues/2192
Fixes: https://github.com/aws/aws-sdk-js-v3/issues/2199
Fixes: https://github.com/aws/aws-sdk-js-v3/issues/2219

### Description
This PR marks react-native-get-random-values as optional peer dependency to prevent install on non React Native environments.

### Testing
Verified that ReactNative E2E tests are successful.

<details>
<summary>Command line output</summary>

```console
$ pwd
/Users/trivikr/workspace/aws-sdk-js-v3/tests/react-native/End2End

$ node launch-app.js --local-publish
...
 PASS  e2e/End2EndTest.spec.js (14.5s)
  S3
    ✓ should upload an object after tapping putObject button (2276ms)
    ✓ after tapping getObject, App should get the same object that uploaded (2224ms)
    ✓ should list Objects after tapping listObjects button (1496ms)
    ✓ should conduct multipart upload after tapping createMultipartUpload, uploadPart, completeMultipartUpload buttons (3096ms)
```

</details>

<details>
<summary>Details</summary>

https://user-images.githubusercontent.com/16024985/112871068-bb1fe300-9073-11eb-9d25-13d0a2f67ab0.mov

</details>

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

aws-sdk-js-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: sdk-staging-test
* Commit ID: 00a8e206758c1133bce8e10a49cce41d0d473440
* Result: SUCCEEDED
* [Build Logs](https://09vzxhs6t5.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=904b2924-513f-4b73-a89c-57f872463e42%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

G-Rath commented:
@trivikr have you tested this to confirm it doesn't throw an error on import if the package isn't installed?

I would expect it to, which is why we used a try/catch over in #2108

trivikr commented:
> @trivikr have you tested this to confirm it doesn't throw an error on import if the package isn't installed?

The AWS SDK for JavaScript no longer imports react-native polyfills, and expect them to be imported by the app using the SDK in react-native environment.

I'll evaluate the behavior if the react-native app doesn't contain react-native polyfills, and update this PR.

G-Rath commented:
Ah right I see - sorry it didn't click that [this change](https://github.com/aws/aws-sdk-js-v3/pull/2108/files#diff-97aa9e56ffa74985e40882bbd10c295acb15aa104f690bab84de3da37de41a81R2-R7) was an *add*, not a modify 😅

trivikr commented:
The reason for removing require/import for `react-native-get-random-values` dependency is that our react-native End2End tests fail with "Unable to resolve module" error.

<details>
<summary>Screenshot</summary>

<img width="559" alt="Unable-to-resolve-module-error" src="https://user-images.githubusercontent.com/16024985/112908489-752f4300-90a4-11eb-9265-8172a907b7c7.png">

</details>

Just like uuid expects it's users to import react-native polyfills, AWS SDK for JavaScript would do the same.

trivikr commented:
This PR is currently pending review from amplify-js team.
cc @Ashish-Nanda @amhinson

trivikr commented:
Notes from the meeting on 3/31 with Amplify team:
* Remove requirement of react-native polyfills as peerDependencies, as peerDependenciesMeta just removes the warnings but installs peerDependencies by default.
* Check if a custom error can be emitted for react-native customers requesting particular polyfill as a requirement for aws-sdk just like uuid does https://github.com/uuidjs/uuid#getrandomvalues-not-supported. Amplify team had tried in the past, and it was not straight-forward.

aws-sdk-js-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: sdk-staging-test
* Commit ID: 31a2f8bd4da97f02445c7b1bc69be7a948d40a25
* Result: SUCCEEDED
* [Build Logs](https://09vzxhs6t5.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=638d6b78-d770-4606-b2e4-da28d78fe0f5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

trivikr commented:
> * Remove requirement of react-native polyfills as peerDependencies, as peerDependenciesMeta just removes the warnings but installs peerDependencies by default.

Removed in [`5baa663` (#2191)](https://github.com/aws/aws-sdk-js-v3/pull/2191/commits/5baa663adb6b9190c70dc53efd6e0f20d7374c40)

> * Check if a custom error can be emitted for react-native customers requesting particular polyfill as a requirement for aws-sdk just like uuid does [uuidjs/uuid#getrandomvalues-not-supported](https://github.com/uuidjs/uuid#getrandomvalues-not-supported). Amplify team had tried in the past, and it was not straight-forward.

This isn't possible, as it's an implementation detail in https://github.com/uuidjs/uuid/blob/91805f665c38b691ac2cbda56a99231432b00a1a/src/rng-browser.js#L21-L25

I've updated README to share react-native polyfills import example instead [`d67256e` (#2191)](https://github.com/aws/aws-sdk-js-v3/pull/2191/commits/d67256e9d1d07a3e74f3b301efd11f57dadf95ed)

aws-sdk-js-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: sdk-staging-test
* Commit ID: d67256e9d1d07a3e74f3b301efd11f57dadf95ed
* Result: SUCCEEDED
* [Build Logs](https://09vzxhs6t5.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=69c16298-2442-4c8c-b516-156f5a6b1c1b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-sdk-js-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: sdk-staging-test
* Commit ID: 751603855c519728c3b12e521eadff2b172f1faf
* Result: SUCCEEDED
* [Build Logs](https://09vzxhs6t5.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=1042531d-698d-4571-ad39-5f68f30b71bf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-sdk-js-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: sdk-staging-test
* Commit ID: d3fa8a130cdd2b8a3e285e8a38a198516e98e43f
* Result: SUCCEEDED
* [Build Logs](https://09vzxhs6t5.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=dbeff1fb-ea1f-4e6d-bd6a-c1b810259ffc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

iartemiev reviewed (approved):
Approved after testing the effect of this PR on Amplify RN apps.

AllanZhengYP reviewed (approved):
So nice to get rid of react-native from most users' dependencies.

J-Krush commented:
So much appreciation for this PR! Unblocked on my work now :D
